core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -135,18 +130,52 @@ else if (partitioning.equals(SCALED_WRITER_DISTRIBUTION)) {
                     physicalWrittenBytesSupplier,
                     writerMinSize);
         }
+        else if (partitioning.equals(SCALED_WRITER_HASH_DISTRIBUTION) || isScalingPartitioning(partitioning)) {
+            exchangerSupplier = () -> {
+                int bucketCount = getTaskScaleWritersPartitionCount(session);
+
+                PartitioningHandle unwrappedPartitioning = partitioning;
+                if (isScalingPartitioning(partitioning)) {
+                    unwrappedPartitioning = unwrapScalingPartitioningHandle(partitioning);
+                }
+
+                PartitionFunction partitionFunction = partitionFunctionFactory.create(
+                        session,
+                        unwrappedPartitioning,
+                        partitionChannels,
+                        partitionChannelTypes,
+                        partitionHashChannel,
+                        bucketCount);
+                Function<Page, Page> partitionPagePreparer;
+                if (isSystemPartitioning(unwrappedPartitioning)) {
```

Comment:
Consider extracting this to a method (e.g. `createPartitionPagePreparer`) and reusing it in the following `if`